consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -26,16 +26,53 @@ use commonware_runtime::{
     buffer::PoolRef, spawn_cell, Clock, ContextCell, Handle, Metrics, Spawner, Storage,
 };
 use commonware_storage::journal::segmented::variable::{Config as JConfig, Journal};
+use commonware_utils::futures::AbortablePool;
+use core::{future::Future, panic};
 use futures::{
     channel::{mpsc, oneshot},
-    future::Either,
     pin_mut, StreamExt,
 };
 use prometheus_client::metrics::{counter::Counter, family::Family, histogram::Histogram};
 use rand::{CryptoRng, Rng};
-use std::num::NonZeroUsize;
+use std::{
+    num::NonZeroUsize,
+    pin::Pin,
+    task::{self, Poll},
+};
 use tracing::{debug, info, trace, warn};
 
+/// A outstanding request to the automaton.
+struct Request<D: Digest, P: PublicKey, R>(
+    /// Attached context for the pending item.
+    Context<D, P>,
+    /// Oneshot receiver that the automaton is expected to respond over.
+    oneshot::Receiver<R>,
+);
+
+impl<D: Digest, P: PublicKey, R> Request<D, P, R> {
+    pub fn is_for_view(&self, view: View) -> bool {
+        self.0.view() == view
+    }
+}
+
+/// Adapter that polls an [Option<Request<D, P, R>>] in place.
+struct Waiter<'a, D: Digest, P: PublicKey, R>(&'a mut Option<Request<D, P, R>>);
+
+impl<'a, D: Digest, P: PublicKey, R> Future for Waiter<'a, D, P, R> {
+    type Output = (Context<D, P>, Result<R, oneshot::Canceled>);
+
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let Waiter(slot) = self.get_mut();
+        match slot.as_mut() {
+            None => Poll::Pending,
+            Some(Request(ctx, ref mut receiver)) => match Pin::new(receiver).poll(cx) {
+                Poll::Ready(res) => Poll::Ready((ctx.clone(), res)),
```

Comment:
I guess we have to clone() again here because we only get a mutable reference to cx?

Reply:
Lemme see if it can be done without a clone